### DISCUSSION TITLE
Couldn't compile Embarcadero C++ Builder for a long time

### DISCUSSION THREAD

**Discussion by vadim10012** [harfbuzz/harfbuzz#3618] [Q&A] [answered]
Hello. 

I decided to compile harfbuzz on Embarcadero Rad Studio 11.1 C++ Builder clang compiler and lost a lot of time.
In the source:
1.  hb-algs.hh , line 180 :  
    _hb_roundf (float x) { return floorf (x + .5f); }
    you need to **change floorf to floor** and add #include "math.h"
2.   hb-map.hh , line 278
  auto keys () const HB_AUTO_RETURN
  (
    + hb_array (items, mask ? mask + 1 : 0)
    | hb_filter (&item_t::is_real)
    | hb_map (&item_t::key)
**//rem    | hb_map (hb_ridentity)**
  ) 
I had to comment out this line otherwise the hb-subset.c source would not be compiled. I'm not good with templates, so I couldn't fix the source.

Could you fix the sources so people don't spend a lot of time compiling in C++ Builder clang compiler?

**Comment by behdad** [comment, marked as answer]
Hello. I didn't see this before, sorry about that.

> you need to change floorf to floor and add #include "math.h"

But we include `cmath`, which is the correct way to include `math.h` in C++. Also `floorf` is the correct way to floor a `float` value as opposed to a `double`, which is what `floor` does.

> //rem | hb_map (hb_ridentity)

That line is necessary to convert a reference to an rvalue and is intentional.

I'm sorry you had to spend time changing the source code, but these seem to be shortcomings of the compiler you are using, not problems in the source code.
